Oauth grant classes cleanup, added refresh token request

lightoze opened the pull request:
This patch does some cleanup in OAuth grant classes and also ground work for refresh token handling.

jameskleeh commented:
There are a few issues here:

* Why was AuthorizationCodeGrant and PasswordGrant changed?
  * Passing the client id and secret for authentication is a valid scheme and must be supported
* You can't change the public variable name in ClientCredentialsGrant. That is a breaking change.
* Is there a specification that says multiple scopes should be space separated for the client credentials grant? 
* OpenIdRefreshTokenRequestContext, OauthRefreshTokenRequestContext
  * Need the @author and @since tags modified to represent reality. Java doc needs updated

lightoze commented:
> * Why was AuthorizationCodeGrant and PasswordGrant changed?
>   * Passing the client id and secret for authentication is a valid scheme and must be supported

I removed code which was not used and didn't seem logical. Due to form serialization grant classes are not suitable as is for using them as request object so you have to convert them to `SecureGrantMap`. In terms of `TokenRequestContext`, that map **is** the grant, so **it** should implement `SecureGrant`.

> * You can't change the public variable name in ClientCredentialsGrant. That is a breaking change.

Valid point, but I doubt anyone is using it. Also I just now noticed that its `toMap` method returns normal hash map instead of `SecureGrantMap` which means client id/secret won't be sent, so if anyone would attempt using it it likely wouldn't work.

> * Is there a specification that says multiple scopes should be space separated for the client credentials grant?

Yes. https://tools.ietf.org/html/rfc6749#section-4.4.2

> * OpenIdRefreshTokenRequestContext, OauthRefreshTokenRequestContext
>   * Need the @author and @SInCE tags modified to represent reality. Java doc needs updated

Will fix. Should it be `@since 1.3`? I'll wait for your replies about all topics and make an update.

jameskleeh commented:
You may be right about the id/secret not working in its current form, however that does not mean the code should be removed without replacement. Yes this should be `@since 1.3.0`

lightoze commented:
Well the code that cannot and should not be used :)

lightoze commented:
I've made changes as discussed except for `ClientCredentialsGrant` - after reading the code it was broken at so many points (wrong value of constant in question, not returning `SecureMap` and thus unable to use client authentication) that I instead made even more incompatible change and changed constants to be private - in line with other grant types. I hope you understand.

`SecureGrant` is now implemented by new abstract class. When micronaut-projects/micronaut-core#1853 is fixed, all `toMap` methods can be deprecated and grant objects can be used directly.

lightoze commented:
@jameskleeh Is there anything yet to be resolved before this can be merged?

jameskleeh commented on security-oauth2/src/main/java/io/micronaut/security/oauth2/endpoint/token/request/context/OpenIdCodeTokenRequestContext.java:
```diff
@@ -56,19 +53,6 @@ public OpenIdCodeTokenRequestContext(AuthorizationResponse authorizationResponse
-     * @param clientConfiguration The client configuration
-     * @return The media type
-     */
-    protected static MediaType getMediaType(OauthClientConfiguration clientConfiguration) {
```
Moving this to `AbstractTokenRequestContext` doesn't make sense because this logic is specific to openid

lightoze replied:
Fixed in new commit.

jameskleeh commented on security-oauth2/src/main/java/io/micronaut/security/oauth2/grants/RefreshTokenGrant.java:
```diff
@@ -32,7 +35,11 @@
 @Introspected
 @JsonNaming(PropertyNamingStrategy.SnakeCaseStrategy.class)
-public class RefreshTokenGrant {
+public class RefreshTokenGrant extends AbstractSecureGrant implements AsMap {
```
Where does the spec say that the refresh token grant needs secured?

lightoze replied:
https://tools.ietf.org/html/rfc6749#section-6
>    Because refresh tokens are typically long-lasting credentials used to
   request additional access tokens, the refresh token is bound to the
   client to which it was issued.  If the client type is confidential or
   the client was issued client credentials (or assigned other
   authentication requirements), the client MUST authenticate with the
   authorization server as described in Section 3.2.1.

jameskleeh commented:
@lightoze See my comments

lightoze commented:
@jameskleeh I consolidated previous changes and rebased everything onto master. Should be good to merge before 2.0.0.
